### PR TITLE
DAOS-5313 container: add retry for snapshot fetch

### PR DESCRIPTION
**Pull request by wangdi1**
Remove temporay maxim snapshot limitation, and add
retry for snapshot IV fetch if the fetching buffer
is not big enough.

Add ifo_misc for cart iv fetch so it will return size
information for REC2BIG error.

Move some crt_proc_d_sg_list_t from object module to
cart so ifo_misc can use it.

Add tests in rebuild_simple to verify the large
snapshot handling.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Comment by wangdi1** [comment]
Alex: could you please check IV cart changes ? Thanks

File: src/cart/crt_iv.c [hunk 1]
```diff
@@ -1418,6 +1419,7 @@ crt_hdlr_iv_fetch_aux(void *arg)
 			D_GOTO(send_error, rc);
 		}
 	} else {
+		output->ifo_misc = iv_value;
```
Comment (frostedcmos) [inline src/cart/crt_iv.c:1422]: i dont think we should be overriding iv_value like this. instead i was thinking ivo_on_fetch() should have new argument that it can return 'misc info' which it can populate on failure if it wants and which would be assigned here.
If we fail for some reason in ivo_on_get/put parts and get here, iv_value will be not what you expect.

File: src/cart/crt_iv.c [hunk 2]
```diff
@@ -410,20 +410,21 @@ crt_ivf_finalize(struct iv_fetch_cb_info *iv_info, crt_iv_key_t *iv_key,
...
 	} else {
 		iv_info->ifc_comp_cb(iv_info->ifc_ivns_internal,
-					iv_info->ifc_class_id,
-					iv_key, NULL,
-					iv_value,
-					output_rc,
-					iv_info->ifc_comp_cb_arg);
+				     iv_info->ifc_class_id,
+				     iv_key, NULL,
+				     output_rc == 0 ? iv_value : output_value,
```
Comment (frostedcmos) [inline src/cart/crt_iv.c:425]: i dont think we should overload meaning of iv_value here. instead i was thinking ifc_comp_cb() should be expanded and along with output_rc, it would contain optional output_value that user can interpreset as they wish

File: src/cart/crt_iv.c [hunk 3]
```diff
@@ -410,20 +410,21 @@ crt_ivf_finalize(struct iv_fetch_cb_info *iv_info, crt_iv_key_t *iv_key,
 
 			output = crt_reply_get(rpc);
 			output->ifo_rc = output_rc;
-
+			output->ifo_misc = *output_value;
 			rc = crt_reply_send(rpc);
+			memset(&output->ifo_misc, 0, sizeof(output->ifo_misc));
```
Comment (frostedcmos) [inline src/cart/crt_iv.c:415]: not sure why you do memsets of those in number of places. those shouldnt be needed

File: src/cart/crt_iv.c [hunk 4]
```diff
@@ -1628,7 +1632,7 @@ crt_iv_fetch(crt_iv_namespace_t ivns, uint32_t class_id,
 		iv_ops->ivo_on_refresh(ivns_internal, iv_key, 0,
 				NULL, false, rc, user_priv);
 		fetch_comp_cb(ivns_internal, class_id, iv_key, NULL,
-			      NULL, rc, cb_arg);
+			      iv_value, rc, cb_arg);
```
Comment (frostedcmos) [inline src/cart/crt_iv.c:1635]: same comment as above. instead of using iv_value as 'misc_value' in error case, a new arg to fetch_comp_cb() should be added instead

**Review by frostedcmos — DISMISSED** [review]

File: src/cart/crt_iv.c (hunk 1 above)
Comment (wangdi1) [reply src/cart/crt_iv.c:1422]: Actually the user only check iv_value for certain failure. as long as on_get/put does not return the same failure. it might be ok.

But this is not good indeed. hmm.

File: src/cart/crt_iv.c (hunk 2 above)
Comment (wangdi1) [reply src/cart/crt_iv.c:425]: hmm, ok. let me try.

File: src/cart/crt_iv.c (hunk 3 above)
Comment (wangdi1) [reply src/cart/crt_iv.c:415]: oh, this is iv_value, if I do not reset it here, then RPC destroy will try to destory the value here.

Comment (frostedcmos) [reply src/cart/crt_iv.c:415]: ugh yeah i see, this feels very hacky; probably more reasons to not override iv_value to avoid situations like these

File: src/container/container_iv.c [hunk 5]
```diff
@@ -699,21 +701,35 @@ int
...
+	D_ALLOC(iv_output, iv_output_size);
+	if (iv_output == NULL)
 		return -DER_NOMEM;
-
-	rc = cont_iv_fetch(ns, IV_CONT_SNAP, cont_uuid, iv_entry,
-			   iv_entry_size, true);
-	if (rc)
+retry:
+	rc = cont_iv_fetch(ns, IV_CONT_SNAP, cont_uuid, iv_output,
+			   iv_output_size, iv_output_size, true);
```
Comment (NiuYawei) [inline src/container/container_iv.c:715]: Sorry, it's bit hard to follow the 'output size' and 'entry size', they are supposed to be same here?

**Review by NiuYawei — COMMENTED** [review]
I just reviewed the changes in container IV and server IV, before proceeding to pool IV changes, I'd like to know why don't we solve this common "receiving buffer not big enough" problem in generic IV framework? Do we have to duplicate the code in each type of IV entries (cont IV, pool IV, etc.)?

File: src/container/container_iv.c [hunk 6]
```diff
@@ -205,28 +176,42 @@ cont_iv_ent_destroy(d_sg_list_t *sgl)
...
+		if (dst_iv_buf_size < size) {
+			/* The caller will retry with bigger buffer size */
+			D_DEBUG(DB_MD, "%zd < %zd\n", dst_iv_buf_size, size);
+			iv_output->iv_entry.iv_result.size = size;
+			iv_output->iv_status = -DER_REC2BIG;
+			return 0;
+		}
+		size = offsetof(struct cont_iv_snapshot,
+				snaps[src->iv_snap.snap_cnt]);
+		memcpy(&dst->iv_snap, &src->iv_snap, size);
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:191]: oh, the new patch does not use iv_status anymore, instead it will use -1 to indicate it needs bigger buffer.

File: src/object/srv_obj.c [hunk 7]
```diff
@@ -1623,24 +1623,21 @@ obj_ioc_init(uuid_t pool_uuid, uuid_t coh_uuid, uuid_t cont_uuid, int opc,
...
+	} else {
+		/**
+		 * The server handle is a dummy and never attached by
+		 * a real container
+		 */
+		D_DEBUG(DB_TRACE, DF_UUID"/%p is server cont hdl\n",
+			DP_UUID(coh_uuid), coh);
 	}
 
-	if (!is_container_from_srv(pool_uuid, coh_uuid)) {
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/object/srv_obj.c:1624]: oh, we already put server-server (pool/container) handle inside the local handle cache by IV when start the pool service. So we do not need it anymore, besides this is also wrong to access the IV in a non-0 xstream.

File: src/container/container_iv.c [hunk 8]
```diff
@@ -205,35 +156,58 @@ cont_iv_ent_destroy(d_sg_list_t *sgl)
...
+		}
+
+		D_DEBUG(DB_MD, "snap_cnt "DF_U64":"DF_U64"\n",
+			snap_cnt, src->iv_snap.snap_cnt);
+		size = cont_iv_snap_ent_size(snap_cnt);
+		if (size > dst_sgl->sg_iovs[0].iov_buf_len) {
+			/* Return -1 so the client fetch will reallocate
+			 * the buffer.*/
+			dst->iv_snap.snap_cnt = (uint64_t)-1;
+			dst->iv_snap.snaps[0] = src->iv_snap.snap_cnt;
```
Comment (NiuYawei) [inline src/container/container_iv.c:184]: I don't see why we have to set -1 to 'snap_cnt', why don't directly set the requested number in 'snap_cnt'?

File: src/pool/srv_iv.c [hunk 9]
```diff
@@ -368,21 +369,60 @@ pool_iv_conn_insert(struct pool_iv_conns *conns, struct pool_iv_conn *new_conn)
...
-	sgl->sg_iovs[0].iov_buf = new_conns;
-	sgl->sg_iovs[0].iov_buf_len = new_size;
-	return 0;
+	dst_entry = dst_sgl->sg_iovs[0].iov_buf;
+	dst_conns = &dst_entry->piv_conn_hdls;
+	dst_conns_size = dst_conns->pic_size + src_conns->pic_size;
+	if (dst_conns_size > dst_conns->pic_buf_size) {
+		D_DEBUG(DB_MD, "dst_conns_size %d > pic_buf_size %u\n",
+			dst_conns_size, dst_conns->pic_buf_size);
+		dst_conns->pic_size = (uint32_t)(-1);
```
Comment (NiuYawei) [inline src/pool/srv_iv.c:392]: Same here, why don't directly set the requested 'dst_conns_size" to pic_size?

File: src/container/container_iv.c (hunk 8 above)
Comment (wangdi1) [reply src/container/container_iv.c:184]: It needs a way to tell if the reply is for requesting more snapshot buffer, or it is just correct reply.  And error code itself can not pack extra information, that is why it use -1 here.

Comment (wangdi1) [reply src/container/container_iv.c:166]: Then that will add rc for every case. even with the value inside iv_entry, which actually does not make things clearer.

File: src/container/container_iv.c [hunk 10]
```diff
@@ -238,7 +211,7 @@ cont_iv_snap_ent_create(struct ds_iv_entry *entry, struct ds_iv_key *key)
 	d_iov_t			key_iov;
 	d_iov_t			val_iov;
 	daos_epoch_t		*snaps = NULL;
-	int			snap_cnt = MAX_SNAP_CNT;
+	int			snap_cnt = -1;
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:214]: ok

File: src/container/container_iv.c [hunk 11]
```diff
@@ -504,61 +461,71 @@ cont_iv_ent_update(struct ds_iv_entry *entry, struct ds_iv_key *key,
...
+					      civ_ent->iv_capa.flags,
+					      civ_ent->iv_capa.sec_capas);
 			if (rc)
-				return rc;
-		} else if (entry->iv_class->iv_class_id == IV_CONT_SNAP) {
-			struct cont_iv_entry *civ_ent = src->sg_iovs[0].iov_buf;
-
+				D_GOTO(out, rc);
+		} else if (entry->iv_class->iv_class_id == IV_CONT_SNAP &&
+			   civ_ent->iv_snap.snap_cnt != (uint64_t)(-1)) {
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:477]: oh, -1 is possible for update path as well. So during iv fetch, if the root return -1, then  the intermediate node will refresh the local node (ivc_refresh callback), and then return to the caller.

File: src/container/container_iv.c [hunk 12]
```diff
@@ -772,53 +734,52 @@ cont_iv_snapshots_update(void *ns, uuid_t cont_uuid, uint64_t *snapshots,
...
+	uint64_t		snap_cnt = INIT_SNAP_CNT;
+	int			rc;
 
 	/* Only happens on xstream 0 */
 	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+retry:
+	entry_size = cont_iv_snap_ent_size(snap_cnt);
+	rc = cont_iv_fetch(ns, IV_CONT_SNAP, cont_uuid, &iv_entry,
+			   sizeof(iv_entry), entry_size, false /* retry */);
+	if (rc)
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:774]: retryable is checked inside iv_op().  or I miss the question.
Retry is only needed based on different IV cases and caller use cases. snapshot fetch does not need it, Or I probably miss your question?

File: src/container/container_iv.c [hunk 13]
```diff
@@ -945,9 +907,8 @@ cont_iv_ec_agg_eph_update_internal(void *ns, uuid_t cont_uuid,
 	if (rc)
 		return rc;
 
-	rc = cont_iv_update(ns, op, cont_uuid, &iv_entry,
-			    sizeof(struct cont_iv_entry), shortcut,
-			    sync_mode, false /* retry */);
+	rc = cont_iv_update(ns, op, cont_uuid, &iv_entry, sizeof(iv_entry),
+			    shortcut, sync_mode, false /* retry */);
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:911]: These retry will only happen in the caller side, and the middle node just forward these failure to the caller anyway.

File: src/iosrv/server_iv.c [hunk 14]
```diff
@@ -406,9 +390,21 @@ ivc_on_fetch(crt_iv_namespace_t ivns, crt_iv_key_t *iv_key,
...
-	rc = fetch_iv_value(entry, &key, iv_value, &entry->iv_value, priv);
-	if (rc == 0)
-		entry->iv_valid = true;
+	if (entry->iv_class->iv_class_ops &&
+	    entry->iv_class->iv_class_ops->ivc_ent_fetch)
+		rc = entry->iv_class->iv_class_ops->ivc_ent_fetch(entry, &key,
+								  iv_value,
+								  priv);
+	else
+		rc = daos_sgl_copy_data(iv_value, &entry->iv_value);
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/iosrv/server_iv.c:399]: ivc_ent_fetch() could be NULL

File: src/pool/srv_iv.c [hunk 15]
```diff
@@ -347,27 +348,138 @@ pool_iv_conn_insert(struct pool_iv_conns *conns, struct pool_iv_conn *new_conn)
...
+	int			rc = 0;
+
+	/* Insert all connections from src_conns to dst_sgl */
+	conn = src_conns->pic_conns;
+	end = (char *)conn + src_conns->pic_size;
+	while (pool_iv_conn_valid(conn, end)) {
+		int ret;
+
+		ret = pool_iv_conn_insert(dst_conns, conn);
+		rc = rc == 0 ? ret : rc;
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/pool/srv_iv.c:371]: ah, it should indeed. initially, I want to skip the DER_NONEXIT case, but I change the pool_iv_conn_insert().

File: src/pool/srv_iv.c [hunk 16]
```diff
@@ -347,27 +348,138 @@ pool_iv_conn_insert(struct pool_iv_conns *conns, struct pool_iv_conn *new_conn)
...
+	struct pool_iv_conns	*dst_conns;
+	uint32_t		dst_conns_size;
+	int			rc = 0;
+
+	src_conns = &src_iv->piv_conn_hdls;
+	D_ASSERT(src_conns->pic_size != (uint64_t)(-1));
+	if (src_conns->pic_size == 0)
+		return 0;
+
+	if (src_conns->pic_size == (uint32_t)(-1)) {
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/pool/srv_iv.c:394]: sigh, that ASSERT should be removed. Thanks

File: src/pool/srv_iv.c [hunk 17]
```diff
@@ -347,27 +348,138 @@ pool_iv_conn_insert(struct pool_iv_conns *conns, struct pool_iv_conn *new_conn)
...
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/pool/srv_iv.c:453]: That assert in 449 should be removed.

**Comment by wangdi1** [comment]
Reviewers, please review this patch. Thank you!

**Review by liuxuezhao — APPROVED** [review]
still one issue to confirm for retry parameter, and seems need rebase...

**Review by NiuYawei — APPROVED** [review]

**Comment by ashleypittman** [comment]
This got stuck in testing on what looks like an OSA test which has since been disabled in master, if you merge then that issue should be resolved.  The PR also has conflicts so you'll need to merge before landing anyway.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Need to figure out why it hung CI but at the same time, it has a conflict against latest master

**Comment by wangdi1** [comment]
Just resolve the conflict, and please review. Thanks.

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — DISMISSED** [review]

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Looks like some aggregation tests are failing...maybe timing related as they are checking that space is freed?

**Comment by wangdi1** [comment]
resolve the conflict and fix the CI test failure. Please review. Thanks.

**Comment by wangdi1** [comment]
Reviewers,

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]
